encode.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2018 gopcua authors. All rights reserved.
+// Use of this source code is governed by a MIT-style license that can be
+// found in the LICENSE file.
+
+package gopcua
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"reflect"
+	"time"
+
+	"github.com/wmnsk/gopcua/utils"
+)
+
+const (
+	null    = 0xffffffff
+	f32qnan = 0xffc00000
+	f64qnan = 0xfff8000000000000
+)
+
+// BinaryDecoder is the interface implemented by an object that can
+// unmarshal a binary OPC/UA representation of itself.
+type BinaryDecoder interface {
+	DecodeOPCUA(io.Reader) error
+}
+
+// BinaryEncoder is the interface implemented by an object that can
+// marshal itself into a binary OPC/UA representation.
+type BinaryEncoder interface {
+	EncodeOPCUA(io.Writer) error
+}
+
+// Bytes is a byte array without length encoding which can be
+// marshalled into a binary OPC/UA representation.
+//
+// It must be the last element in a struct since it consumes
+// all remaining bytes of the input stream.
+type Bytes []byte
+
+func (v Bytes) EncodeOPCUA(w io.Writer) error {
+	_, err := w.Write([]byte(v))
+	return err
+}
+
+func (v *Bytes) DecodeOPCUA(r io.Reader) error {
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	*v = make([]byte, len(b))
```

Comment:
couldn't this be replaced with just:
```go
*v = Bytes(b)
```
?
this removes one possibly big allocation.

Reply:
I have refactored that code a bit and dropped the `Bytes` type when I started to integrate this with the existing code. You can now tag a `[]byte` with `opcua:"raw"` to just read the raw bytes and there is a fast path in the codec for that case. I'll have a look there. I'll push an update a bit later.